src/Ng/Lightning.cs

```diff
@@ -0,0 +1,397 @@
+                            // Fetch graph for package version
+                            var url = line.TrimEnd();
+                            var graph = await collectorHttpClient.GetGraphAsync(new Uri(url));
+                            if (sortedGraphs.ContainsKey(url))
+                            {
+                                sortedGraphs[url] = graph;
+                            }
+                            else
+                            {
+                                sortedGraphs.Add(url, graph);
+                            }
+
+                            // To reduce memory footprint, we're flushing out large registrations
+                            // in very small batches.
+                            if (graph.Nodes.Count() > 3000 && sortedGraphs.Count >= 2)
+                            {
+                                // Process graphs
+                                await ProcessGraphsAsync(packageId, sortedGraphs, storageFactory, baseAddress);
+
+                                // Destroy!
+                                sortedGraphs = new Dictionary<string, IGraph>();
+                            }
+
+                            // Read next line
+                            line = await indexStreamReader.ReadLineAsync();
+                        }
+
+                        // Process graphs
+                        if (sortedGraphs.Any())
+                        {
+                            await ProcessGraphsAsync(packageId, sortedGraphs, storageFactory, baseAddress);
+                        }
+
+                        // Update cursor file so next time we have less work to do
+                        batchStart++;
+                        await UpdateCursorFileAsync(cursorFile, batchStart, batchEnd);
+                    }
+                    catch (Exception)
+                    {
+                        UpdateCursorFileAsync(cursorFile, batchStart, batchEnd).Wait();
+                        throw;
+                    }
+                }
+            }
+
+            await UpdateCursorFileAsync("DONE" + cursorFile, batchStart, batchEnd);
+            log.WriteLine("Finished lightning strike for {0}.", cursorFile);
+        }
+
+        private static Task UpdateCursorFileAsync(string cursorFileName, int startIndex, int endIndex)
+        {
+            using (var streamWriter = new StreamWriter(cursorFileName))
+            {
+                streamWriter.Write(startIndex);
+                streamWriter.Write(",");
+                streamWriter.Write(endIndex);
+            }
+
+            return Task.FromResult(true);
+        }
+
+        private static Task ProcessGraphsAsync(string packageId, IDictionary<string, IGraph> sortedGraphs, StorageFactory storageFactory, string contentBaseAddress)
+        {
+            return RegistrationCatalogCreator.ProcessGraphs(packageId, sortedGraphs, storageFactory, new Uri(contentBaseAddress), 64, 128, CancellationToken.None);
```

Comment:
I think this is a different implementation of Registration creation than is happening in the incremental job. The incremental job uses RegistrationCollector that delegates to RegistrationMaker. I'd rewritten as part of the apiapps push because I got the feedback that this code was hard to understand. However, the code in RegistrationMaker has some inefficiencies - Joel fixed the worst of these before the break.